windows/src/developer/TIKE/compile/CompileKeymanWeb.pas

```diff
@@ -844,47 +844,54 @@ function TCompileKeymanWeb.JavaScript_OutputString(fkp: PFILE_KEY; pwszOutput: P
 
 function TCompileKeymanWeb.JavaScript_Shift(fkp: PFILE_KEY; FMnemonic: Boolean): Integer;
 begin
-  if not FMnemonic then
+  if FMnemonic then
   begin
-    if (fkp.ShiftFlags and KMX_ISVIRTUALKEY) = KMX_ISVIRTUALKEY then
+    if (fkp.ShiftFlags and KMX_VIRTUALCHARKEY) = KMX_VIRTUALCHARKEY then
     begin
-      if IsKeyboardVersion10OrLater then
-      begin
-        // Full chiral modifier and state key support starts with KeymanWeb 10.0
-        Result := fkp.ShiftFlags;
-      end
-      else
-      begin
-        // Non-chiral support only and no support for state keys
-        if (fkp.ShiftFlags and (
-          KMX_LCTRLFLAG or KMX_RCTRLFLAG or KMX_LALTFLAG or KMX_RALTFLAG)) <> 0 then   // I4118
-        begin
-          ReportError(fkp.Line, CWARN_ExtendedShiftFlagsNotSupportedInKeymanWeb, 'Extended shift flags LALT, RALT, LCTRL, RCTRL are not supported in KeymanWeb');
-        end;
+      ReportError(fkp.Line, CERR_VirtualCharacterKeysNotSupportedInKeymanWeb, 'Virtual character keys not currently supported in KeymanWeb');  // I1971   // I4061
+      Exit(0);
+    end;
 
-        if (fkp.ShiftFlags and (
-          KMX_CAPITALFLAG or KMX_NOTCAPITALFLAG or KMX_NUMLOCKFLAG or KMX_NOTNUMLOCKFLAG or KMX_SCROLLFLAG or KMX_NOTSCROLLFLAG)) <> 0 then   // I4118
-        begin
-          ReportError(fkp.Line, CWARN_ExtendedShiftFlagsNotSupportedInKeymanWeb, 'Extended shift flags CAPS and NCAPS are not supported in KeymanWeb');
-        end;
+    if ((fkp.ShiftFlags and KMX_ISVIRTUALKEY) = KMX_ISVIRTUALKEY) and (Ord(fkp.Key) <= 255) then
+    begin
+      // We prohibit K_ keys for mnemonic layouts. We don't block T_ and U_ keys.
```

Comment:
I'm a little confused. I thought mnemonic layouts only use K_ keys.

And does `255` correspond to the last valid K_ code?

Reply:
Mnemonic layouts match on keycap characters, not keys. `K_` keys correspond to physical keys.  See https://help.keyman.com/developer/language/reference/mnemoniclayout for more detail.

All `K_` codes are in the 0-255 range (there are some specials in the 50,000+ range but they are for internal use). `T_` codes start at 256. We should _probably_ define a constant I guess.